ts/src/profiler.ts

```diff
@@ -13,47 +13,387 @@
+ */
+export class Profiler extends common.ServiceObject {
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private profileLabels: {instance: string};
+  private deployment: Deployment;
+  private profileTypes: string[];
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
+  heapProfiler: HeapProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
-    this.config = config;
+    config = common.util.normalizeArguments(null, config);
+    const serviceConfig = {
+      baseUrl: API,
+      scopes: [scope],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(serviceConfig, config), baseUrl: '/'});
 
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
+    this.config = config;
 
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    this.deployment = {
+      projectId: this.config.projectId,
+      target: this.config.serviceContext.service,
+      labels: {
+        zone: this.config.zone,
+        version: this.config.serviceContext.version
+      }
+    };
+
+    this.profileLabels = {instance: this.config.instance};
+
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler = new TimeProfiler(this.config.timeIntervalMicros);
+    }
+    if (!this.config.disableHeap) {
+      this.profileTypes.push(ProfileTypes.Heap);
+      this.heapProfiler = new HeapProfiler(
+          this.config.heapIntervalBytes, this.config.heapMaxStackDepth);
+    }
+  }
+
+  /**
+   * Starts and endless loop to poll profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * profiler server, this problem will be logged at the debug level and
+   * otherwise ignored.
+   * If there is a problem polling profiler server for instructions
+   * on the type of profile created, this problem will be logged and profile
+   * creation will be retried.
+   * This promise will not resolve while profiling is still occurring. This
+   * promise should not ever be resolved or rejected.
+   */
+  start(): Promise<void> {
+    return this.runLoop();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async runLoop(): Promise<void> {
+    const delayMillis = await this.collectProfile();
+
+    // Schedule the next profile.
+    setTimeout(this.runLoop.bind(this), delayMillis).unref();
+  }
+
+  /**
+   * Waits for profiler server to tell it to collect a profile, then collects
+   * a profile and uploads it.
+   * Returns time, in ms, to wait before asking profiler server if one should
```

Comment:
Use `@return` annotation rather than free form comment to describe the return value.